tools/autograd/templates/variable_factories.h

```diff
@@ -19,6 +20,134 @@ using at::DimnameList;
+      TORCH_CHECK(init_list.size() > 0, "Empty init-list is not supported");
+      scalar_type_ = init_list.begin()->scalar_type_;
+      const ListInitTensor& first_elem = *(init_list.begin());
+      for (const auto& elem : init_list) {
+        TORCH_CHECK(elem.scalar_type_ == first_elem.scalar_type_,
+          "Expected all elements of the tensor to have the same scalar type: ",
+          first_elem.scalar_type_,
+          ", but got element of scalar type: ",
+          elem.scalar_type_);
+        TORCH_CHECK(elem.sizes_ == first_elem.sizes_,
+          "Expected all sub-lists to have sizes: ",
+          first_elem.sizes_,
+          " (e.g. ", first_elem, "), ",
+          "but got sub-list ",
+          elem,
+          " with sizes: ",
+          elem.sizes_);
+      }
+      sizes_.reserve(first_elem.sizes_.size() + 1);
+      sizes_.push_back(init_list.size());
+      sizes_.insert(sizes_.end(), first_elem.sizes_.begin(), first_elem.sizes_.end());
+    }
+
+    const c10::Scalar& scalar() const {
+      return scalar_;
+    }
+
+    const std::initializer_list<ListInitTensor>& init_list() const {
+      return init_list_;
+    }
+
+    const std::vector<int64_t>& sizes() const {
+      return sizes_;
+    }
+
+    const c10::ScalarType& scalar_type() const {
+      return scalar_type_;
+    }
+
+    const ListInitTensorType& type() const {
+      return type_;
+    }
+
+    at::Tensor to_tensor(const at::TensorOptions& options) const {
+      // NOTE: Here we explicitly choose to initialize the tensor on CPU first,
+      // fill each element of the tensor, and then move the tensor to the desired
+      // device. For CUDA device, this approach only involves 1 CUDA kernel launch,
+      // and is much faster than initializing the tensor on CUDA first and then
+      // filling each element of it (which involves `N` CUDA kernel launches where
+      // `N` is the number of the elements in the tensor).
+      at::Tensor tensor = ([&]() {
+        at::AutoNonVariableTypeMode non_var_type_mode(true);
+        return at::empty(sizes_, at::TensorOptions(options).device(at::kCPU).is_variable(false));
+      })();
+      fill_tensor(tensor);
+      return tensor.to(options.device());
+    }
+
+    void pretty_print_recursive(std::ostream& stream) const {
+      if (type_ == ListInitTensorType::Scalar) {
+        AT_DISPATCH_ALL_TYPES_AND3(at::kBool, at::kHalf, at::kBFloat16, scalar_type_, "ListInitTensor_pretty_print_scalar", [&] {
+          stream << const_cast<c10::Scalar&>(scalar_).to<scalar_t>();
+        });
+      } else if (type_ == ListInitTensorType::InitList) {
+        stream << "{";
+        for (const ListInitTensor* it = init_list_.begin(); it != init_list_.end(); it++) {
+          it->pretty_print_recursive(stream);
+          if (std::next(it) != init_list_.end()) stream << ", ";
+        }
+        stream << "}";
+      }
+    }
+
+   private:
+    void fill_tensor(at::Tensor tensor) const {
+      int index = 0;
```

Comment:
nit: You might want to make this size_t (I'm guessing that is the size_type of std::initializer_list). But I don't think anyone will try to create a tensor using 2^32 numbers, so this is probably okay...